### PR TITLE
Type-directed optional fields

### PR DESCRIPTION
**Pull request by friedbrice**
This PR makes a backwards-compatible change to `ToJSON` and `FromJSON` that allows the instance to configure the optionality of record fields of that type when using Generic deriving or Template Haskell deriving.

Specifically, we make the following use-facing changes:

```haskell
class ToJSON a where
    ...
    omitField :: a -> Bool
    omitField _ = False
```

and

```haskell
class FromJSON a where
    ...
    omittedField :: Maybe a
    omittedField = Nothing
```

The new methods have default implementations that preserve the current behavior of Generically-derived and Template Haskell created instances.
Instance authors can use these methods to configure how to represent fields of this type as omitted from a JSON record.
As an added benefit, this PR eliminates the needs for special-case handling of `Maybe` and `Option`, significantly simplifying the Generic and Template Haskell code.

This PR solves https://github.com/haskell/aeson/issues/646 and https://github.com/haskell/aeson/issues/792, and it's similar to https://github.com/haskell/aeson/pull/839 by @phadej.
Please let me know what you think, if this PR is a suitable addition to Aeson, and if there's anything about it you'd like me to fix or change.

Thanks much!

### CODE REVIEW

File: tests/UnitTests/OptionalFields.hs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+optionalFields :: [TestTree]
+optionalFields = [omitGenerics, omitTH, proofOfConcept]
+
+
+-- c.f. https://github.com/haskell/aeson/pull/839#issuecomment-782453060
+data POC = POC
+  { x :: Nullable Int -- Field is required, but can be null.
+  , y :: Undefineable Int -- Field is optional, but cannot be null.
+  , z :: NullOrUndefineable Int -- Field is optional, and can be null.
+  }
```
Comment (friedbrice) [inline tests/UnitTests/OptionalFields.hs:20]: These tests verify that the PR supports your desired behavior (https://github.com/haskell/aeson/pull/839#issuecomment-782453060), @bergmark :)

File: src/Data/Aeson/Types/ToJSON.hs [hunk 2]
```diff
@@ -1291,6 +1273,7 @@ instance ToJSON1 Maybe where
 
 instance (ToJSON a) => ToJSON (Maybe a) where
     toJSON = toJSON1
+    omitField = null
     toEncoding = toEncoding1
```
Comment (friedbrice) [inline src/Data/Aeson/Types/ToJSON.hs:1277]: A `Maybe a` will be omitted when (1) `omitNothingFields` is `True`, and (2) when the maybe is empty.

This preserves the current `ToJSON` behavior of `Maybe a` without needing overlapping instances.

File: src/Data/Aeson/Types/FromJSON.hs [hunk 3]
```diff
@@ -1517,6 +1518,7 @@ instance FromJSON1 Maybe where
 
 instance (FromJSON a) => FromJSON (Maybe a) where
     parseJSON = parseJSON1
+    omittedDefault = Just Nothing
```
Comment (friedbrice) [inline src/Data/Aeson/Types/FromJSON.hs:1521]: A field absent in JSON will be populated by `Nothing` in the Haskell representation, preserving the current behavior of `FromJSON` for `Maybe a` fields, without needing overlapping instances.

File: src/Data/Aeson/Types/ToJSON.hs [hunk 4]
```diff
@@ -176,6 +176,9 @@ class GToJSON' enc arity f where
     -- and 'liftToEncoding' (if the @arity@ is 'One').
     gToJSON :: Options -> ToArgs enc arity a -> f a -> enc
 
+    gOmitField :: proxy enc -> proxy' arity -> f a -> Bool
```
Comment (phadej) [inline src/Data/Aeson/Types/ToJSON.hs:179]: Is idea of having `gOmitField` and `gOmittedDefault` to inherit the default value in wrapper types like `newtype`s or singular `data Wrapper = Wrap SomeOtherType` wrappers?

I don't think it's necessary.

It also is (or at least was) to prefer single method `G*` classes so they are inlined better.

TL;DR I'd rather not add generics for `omitField` nad `omittedField`. It will make patch smaller and we won't need to worry about generically derived code performance.

File: src/Data/Aeson/Types/ToJSON.hs [hunk 5]
```diff
@@ -1291,6 +1273,7 @@ instance ToJSON1 Maybe where
 
 instance (ToJSON a) => ToJSON (Maybe a) where
     toJSON = toJSON1
+    omitField = null
```
Comment (phadej) [inline src/Data/Aeson/Types/ToJSON.hs:1276]: `isNothing` (it's the same as `null`, but more direct).

**Review by phadej — COMMENTED** [review]
Looks good. Having `omitField :: a -> Bool` is indeed simpler that what I had in #839.

There probably should be something in `ToJSON1` and `FromJSON1` classes, as e.g. instance for `Fix` relies on them. How `Fix Maybe` should be serialised?

File: src/Data/Aeson/Types/ToJSON.hs (hunk 4 above)
Comment (friedbrice) [reply src/Data/Aeson/Types/ToJSON.hs:180]: Okay, I'll see about removing them :-)

Comment (friedbrice) [reply src/Data/Aeson/Types/ToJSON.hs:181]: Done.

**Comment by friedbrice** [comment]
I was able to get the generic classes down to one method like you suggested. I still need to work on the `ToJSON1` logic.

File: src/Data/Aeson/Types/FromJSON.hs [hunk 6]
```diff
@@ -251,7 +251,10 @@ class GFromJSON arity f where
     -- or 'liftParseJSON' (if the @arity@ is 'One').
     gParseJSON :: Options -> FromArgs arity a -> Value -> Parser (f a)
 
+class GOmittedDefault arity f where
```
Comment (phadej) [inline src/Data/Aeson/Types/FromJSON.hs:254]: I probably was unclear: Don't add `gOmittedDefault` nor `gOmitField`. The generic deriving is cumbersome already. Unwrapping single constructors isn't a huge task to do manually, and I argue it's better to always be explicit which value will be omitted.

**Comment by phadej** [comment]
I marked this as draft.

Please ping when you seem happy again.

File: tests/UnitTests/OptionalFields.hs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+-- c.f. https://github.com/haskell/aeson/pull/839#issuecomment-782453()6()
+data POC = POC
+  { x :: Nullable Int -- Field is required, but can be null.
+  , y :: Undefineable Int -- Field is optional, but cannot be null.
+  , z :: NullOrUndefineable Int -- Field is optional, and can be null.
+  }
+  deriving (Eq, Show, Generic)
+
+instance ToJSON POC where
+  toJSON = genericToJSON opts
```
Comment (phadej) [inline tests/UnitTests/OptionalFields.hs:22]: missing `toEncoding`

**Comment by friedbrice** [comment]
@phadej I'm pretty happy with it, now. I'm interested in your feedback :-)

**Comment by friedbrice** [comment]
Oh, also, @phadej, I do have some ideas about changes to `To`/`FromJSON1`, but I'd like to know more about what you have in mind before attempting that.

**Review by phadej — COMMENTED** [review]
This looks good.

I have further things to do for this, but I think it's easier and fasterfor me to them myself rather than try to explain them.

Could you squash your changes into own commit (and rebase on recent `master`), so I can continue from there.

File: tests/Encoders.hs [hunk 8]
```diff
@@ -296,29 +296,6 @@ gOptionFieldParseJSON = genericParseJSON optsOptionField
 thMaybeFieldToJSON :: MaybeField -> Value
 thMaybeFieldToJSON = $(mkToJSON optsOptionField 'MaybeField)
 
-
---------------------------------------------------------------------------------
--- IncoherentInstancesNeeded
---------------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (friedbrice) [reply tests/Encoders.hs:302]: Thanks for your help and ideas, btw, @parsonsmatt. The TH was trivially-easy, but the Generics threw me for a loop.

File: src/Data/Aeson/Types/ToJSON.hs [hunk 9]
```diff
@@ -317,6 +316,33 @@ class ToJSON a where
     toEncodingList :: [a] -> Encoding
     toEncodingList = listEncoding toEncoding
 
+    -- | Defines when it is acceptable to omit a field of this type from a record?
```
Comment (friedbrice) [inline src/Data/Aeson/Types/ToJSON.hs:319]:
```suggestion
    -- | Defines when it is acceptable to omit a field of this type from a record.
```
